### ISSUE TITLE
go_proto_library: default "compiler" misinterprets importpath value with semicolon package name suffix

### ISSUE THREAD

**Issue by noahdietz** [bazelbuild/rules_go#3235]
### What version of rules_go are you using?

`v0.33.0`

### What version of gazelle are you using?

`v0.24.0`

### What version of Bazel are you using?

`v5.2.0`

### Does this issue reproduce with the latest releases of all the above?

Yes.

### What operating system and processor architecture are you using?

Linux, AMD64

### Any other potentially useful information about your toolchain?

Using the default "compiler" for go_proto_library, not using a custom `go_proto_compiler` target for custom generation options. I realize that is another approach to take, but wondered if the below was an issue or not.

### What did you do?

Supplied a `path;package` format value to the `importpath` attribute of a `go_proto_library` target. For example, `google.golang.org/genproto/googleapis/cloud/secretmanager/v1;secretmanagerpb`

### What did you expect to see?

The generated code in bazel output with directory name `google.golang.org/genproto/googleapis/cloud/secretmanager/v1`.

### What did you see instead?

The generated code in bazel output with directory name `google.golang.org/genproto/googleapis/cloud/secretmanager/v1;secretmanagerpb`.

The goal was to see if we could override the importpath and package name with go_proto_library targets. The generated package name is properly updated, but the resulting directory structure of the generated code incorrectly includes the `;package` section of the value.

I'm not sure if using `path;package` syntax is actually allowed with `importpath`. If it isn't, perhaps lets throw an error? If it is, perhaps we should fix the generated directory structure to trim the `;package` suffix.

**Comment by noahdietz** [comment]
cc @codyoss

**Comment by achew22** [comment]
Last time I tried something like this, it wasn't supported and I went a different route. I don't think anyone has implemented that but would be very happy to help if you think it's a feature we should adopt. I would like to bear in mind the efforts of `rules_proto` to create a generic way of handling protos are coming along. It might be worth talking to them about using this as an opportunity to switch to the "standard" way of handling protos.

WDYT?

**Comment by noahdietz** [comment]
> Last time I tried something like this, it wasn't supported and I went a different route. 

Yeah, I hear you, and we may just do that, but I felt like it was worth a question to y'all :)

> but would be very happy to help if you think it's a feature we should adopt.

To be clear, the feature in question is (properly) supporting the `path;package` format in the `importpath` value where the `path` segement is used as the generated directory structure (excluding the `;package` suffix) _and_ support overriding the generated package using the suffix following the `;` in `;package`. Are we on the same page?

I would understand if the `importpath` was truly only intended for the `path` format, and that the `path;package` format is explicitly disallowed, that's part of what I wanted to clear up. In which case, we'd want to update the code to throw an error for unsupported format :).

> I would like to bear in mind the efforts of rules_proto to create a generic way of handling protos are coming along. It might be worth talking to them about using this as an opportunity to switch to the "standard" way of handling protos.

I'm all for "standard" :) Happy to look into this. Do you have an issue or RFC you could point me at?

**Comment by achew22** [comment]
Fast answers are better than slow ones so here is the quick one:

> I would understand if the importpath was truly only intended for the path format, and that the path;package format is explicitly disallowed, that's part of what I wanted to clear up. In which case, we'd want to update the code to throw an error for unsupported format :).

Yeah, there is a lot of infra built on `importpath` being just the `path` and nothing more. It would probably require a multi-release process to get `rules_go` and `gazelle` to understand a new format. Happy to discuss more if you're interested in this route.

**Comment by achew22** [comment]
Slower answer broken out because I have low confidence in my ability to find random Google Docs links.

> I'm all for "standard" :) Happy to look into this. Do you have an issue or RFC you could point me at?

Yeah, there is a push to do something common for proto described [here](https://docs.google.com/document/d/1dY_jfRvnH8SjRXGIfg8av-vquyWsvIZydXJOywvaR1A/edit). This is the most promising design doc on the topic so far but it seems to be moving at the speed of Google Open Source 😉 (said as a Xoogler who spent many moons fighting the machine).

**Comment by noahdietz** [comment]
@fmeum awesome, good to know! So in the meantime, do y'all think we should "patch" the existing behavior to be opinionated about the format of `importpath` (i.e. throw an error with `path;package` format input), then once those aforementioned improvements land, we evaluate supporting the new format as a new feature?

I think this would be a bit of churn, and I'm not sure how offensive the existing behavior is (allowing `path;package` format but interpreting it slightly incorrectly) to where it motivates us to start disallowing it in the meantime. So that part of me says "let's leave it the way it is, its not that bad, and convert this issue into a feature request to follow the improvements from https://github.com/bazelbuild/bazel/pull/15854#issuecomment-1184618892".

WDYT?

**Comment by achew22** [comment]
Out of curiosity, is there anywhere other than the `option go_package` directive in proto that uses the `path;package` syntax? Maybe we could use them for inspiration?

**Comment by achew22** [comment]
@noahdietz can you point me to another place in the go ecosystem that is using the `path;package` notation? Off the top of my head I can't think of anywhere else it's used, but I'm not as broad in this ecosystem as I once was.

**Comment by noahdietz** [comment]
@achew22 I must correct myself, it isn't a Go-ism, but really a protobuf-go idea.

**Comment by achew22** [comment]
@noahdietz, given that it is a protobuf-go specific concept, I think adding it into rules_go is probably not the right move. I guess we could work on Gazelle to make it possible to parse the `path;package` syntax if it is completely necessary what do you think about that?

**Comment by codyoss** [comment]
I think the important bit here is that there is a way to overwrite the import *and* package. Exposing a package override as a separate attribute would be fine too, but the `path;package` convention does seem convenient since there is a precedence for this formatting style the in ecosystem already.

**Comment by achew22** [comment]
Maybe I'm confused here, but when I think of the `package`, I think of the thing that comes at the top of a `.go` file and looks like:

```
package asdf
```

Are we talking about the same thing?

**Comment by codyoss** [comment]
We are indeed. In Go it is completely valid to have an import-path of `example.com/something/foo` and in that directory have `package bar`. A great example of this is anytime there is a v2 module.

`example.com/sothing/foobar/v2` could have a actual package name of `foobar`.

**Comment by achew22** [comment]
From an admittedly cursory reading of the code, I don't believe `package` is a part of the build graph. I don't think there is even a way for rules_go to change the package name without running `sed -i s/package foo/package bar/`. It doesn't appear to be a part of the linker options. I think the only way to set the package is to change the `package foo` header that's emitted in the `.go` source file.

If the problem is that `importpath` is not being properly set by Gazelle and it needs to split the `option go_package` into `importpath;package` and set the `importpath` attribute in the `BUILD` file to just be the `importpath` component, then I think I understand what next steps would be here. Would you like to proceed on that path?

**Comment by noahdietz** [comment]
Ok, so this is specific to `go_proto_library` in that it _generates_ the Go code (and thus the `go_library`) that projects depend on. `go_library` and how `importpath` or `package` factors into that is out of scope of this issue. Sorry for the confusion. Let me lay it all out (even if you already know much of this) just to be explicit.

So with the protobuf compiler, protoc, and the protobuf-go code generator, the generator will refer to the `option go_package` to determine the `importpath`, and if specified using the syntax `importpath;package` the `package`, of the Go code it is about to generate. Taking an example, with `option go_package = "github.com/noahdietz/foo;bar";` and running `protoc --go_out=. foo.proto` I'd get `./github.com/noahdietz/foo/foo.pb.go` with `package bar` inside of it.

The same could be said for rules_go, with the exact same proto and an `importpath` attribute value of `github.com/noahdietz/foo`, the package in that generated code would be `bar` - because the `go_package` value specified it. 

Now, at this point, I'm not necessarily pushing for one thing or another, but I want to clarify what the behavior should be. But...

I'm guessing that the `go_proto_library` `importpath` attribute is passed directly to the same `importpath` attribute onthe  resulting `go_library` (or related resulting output target). That would make sense to me. Since rules_go, as you've pointed out, doesn't care what the Go `package` is inside of the `.go` file it wraps in a `go_library`, it makes sense to say that the `importpath` on `go_proto_library` doesn't need to care about the `;package` suffix on the `option go_package` in the source proto.

Where I got to the point of opening this issue was when I experimented with setting the `go_proto_library` `importpath` attribute value on an existing target to something like `importpath = "github.com/noahdietz/foo;bazbuz"` (notice that the suffix/package name `;bazbuz` is different from the proto `option go_package = "github.com/noahdietz/foo;bar"` in my earlier example) - interestingly, the resulting generated Go code had a package of `package bazbuz`(!), but the generated code was output in the directory `bazel-bin/.../github.com/noahdietz/foo;bazbuz/foo.pb.go`. Even though the behavior is effectively undefined (as the [documentation](https://github.com/bazelbuild/rules_go/blob/24c3d41068c57d1ee7bf5774391285fd4ac5d189/proto/core.rst#go_proto_library) literally says `importpath must match the import path specified in .proto files using option go_package`), I was surprised to see the `;bazbuz` in the generated file path/directory structure.

So I'm not sure if it is an accidentally half-implemented feature (that the generated `package` adheres to the `;package` suffix in the `importpath` attribute) or a bug that it respects the `;package` suffix at all and doesn't just throw a validation exception on invalid `importpath` attribute format.

FWIW, protobuf-go does have a plugin option that allows one to override the `importpath;package` of the `go_package` in place for specific proto files: `protoc --go_out=. --go_opt=Mfoo.proto=noahdietz.me/go/foo;fooer foo.proto` generates `./noahdietz.me/go/foo/foo.pb.go` with `package fooer`. That is what motivated me to try tweaking the `importpath` attribute of a `go_proto_library` target.

Sorry for the wall of text. Much of this deatil I should've included in the original bug.
